panic in NDM operator: blockdeviceclaim.(*ReconcileBlockDeviceClaim).isDeviceRequestedByThisDeviceClaim(...)

**What steps did you take and what happened:**
I have a crash loop in my npm operator. I think the crashes started while I was replacing disks. To replace the disks, I followed this process for 3 disks sequentially, provided by the helpful folks on the OpenEBS Slack:

- Pull the old disk (simulated hardware failure, and I couldn't attach more disks to my nodes)
- Insert the new disk
- Make a note of the BD that was auto-created
- Create a new BDC for the disk manually (by copying from the old BDC and changing relevant fields)
- Disable the reconciliation on SPC by adding the annotation reconcile.openebs.io/disable: "true" on SPC
- Update the CSP which has old blockdevice name and old dev link with newly claimed blockdevice and its devlink
- Change the corresponding CSP status to Init (which means that we are recreating pool on newly added disk)
- Wait till all the CVR in that pool to become Healthy (status will ReconstructingNewReplica which means that data is rebuilding from scratch)
- Update the entry of old blockdevice name with new blockdevice name in SPC
- Enable the reconciliation on SPC by removing annotations which added in step1
- If old BDC is not automatically removed, we have to remove one finalizer related to pool i.e `storagepoolclaim.openebs.io/finalizer` then other finalizer will be removed by ndm

For the first disk, this process finished normally. However, for the second disk, I noticed that the BD was not set to Claimed automatically. I continued with the third disk, and its BD was also not set to Claimed automatically.

I investigated further and saw the crash loop in the NDM operator:

```
{"level":"info","ts":1591736885.431024,"logger":"ndm-operator","msg":"Go Version: go1.12.7"}
{"level":"info","ts":1591736885.431187,"logger":"ndm-operator","msg":"Go OS/Arch: linux/arm64"}
{"level":"info","ts":1591736885.4312375,"logger":"ndm-operator","msg":"operator-sdk Version: v0.5.0"}
{"level":"info","ts":1591736885.433303,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1591736885.9101493,"logger":"leader","msg":"Found existing lock with my name. I was likely restarted."}
{"level":"info","ts":1591736885.9109,"logger":"leader","msg":"Continuing as the leader."}
{"level":"info","ts":1591736886.1548922,"logger":"ndm-operator","msg":"Installing the components"}
{"level":"info","ts":1591736898.4133675,"logger":"ndm-operator","msg":"Registering Components"}
{"level":"info","ts":1591736898.7814624,"logger":"ndm-operator","msg":"Check if CR has to be upgraded, and perform upgrade"}
{"level":"info","ts":1591736898.950106,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"blockdevice-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1591736898.9518716,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"blockdeviceclaim-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1591736898.9531066,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"disk-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1591736898.9540648,"logger":"ndm-operator","msg":"Starting the ndm-operator..."}
{"level":"info","ts":1591736899.055072,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"disk-controller"}
{"level":"info","ts":1591736899.0553782,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"blockdeviceclaim-controller"}
{"level":"info","ts":1591736899.0550766,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"blockdevice-controller"}
{"level":"info","ts":1591736899.1557183,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"disk-controller","worker count":1}
{"level":"info","ts":1591736899.156399,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"blockdevice-controller","worker count":1}
{"level":"info","ts":1591736899.1564255,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"blockdeviceclaim-controller","worker count":1}
{"level":"info","ts":1591736899.1569605,"logger":"controller_deviceclaim","msg":"Deleting external dependencies for CR:bdc-ebf6da43-64da-4033-8c9b-82cefad4fc0b","Request.Namespace":"openebs","Request.Name":"bdc-ebf6da43-64da-4033-8c9b-82cefad4fc0b"}
E0609 21:08:19.159792       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/home/travis/.gimme/versions/go1.12.7.linux.arm64/src/runtime/panic.go:522
/home/travis/.gimme/versions/go1.12.7.linux.arm64/src/runtime/panic.go:82
/home/travis/.gimme/versions/go1.12.7.linux.arm64/src/runtime/signal_unix.go:390
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:251
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:281
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:211
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:134
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/home/travis/.gimme/versions/go1.12.7.linux.arm64/src/runtime/asm_arm64.s:1128
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xfab538]

goroutine 262 [running]:
github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x104
panic(0x10ef080, 0x1ff5140)
	/home/travis/.gimme/versions/go1.12.7.linux.arm64/src/runtime/panic.go:522 +0x1a0
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).isDeviceRequestedByThisDeviceClaim(...)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:251
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).releaseClaimedBlockDevice(0x400047b4e0, 0x40002eb6c0, 0x14625e0, 0x40003468a0, 0x19, 0x40004dc801)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:281 +0x3c8
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).FinalizerHandling(0x400047b4e0, 0x40002eb6c0, 0x14625e0, 0x40003468a0, 0x0, 0x4000352660)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:211 +0xb8
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).Reconcile(0x400047b4e0, 0x40004dc8f0, 0x7, 0x4000352660, 0x28, 0x2051720, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:134 +0x4ac
github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x40003028c0, 0x40002b2600)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x158
github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x34
github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0x4000348490)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x50
github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x4000348490, 0x3b9aca00, 0x0, 0x1, 0x40000a1e60)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xb8
github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0x4000348490, 0x3b9aca00, 0x40000a1e60)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x48
created by github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x2dc
```

**The output of the following commands will help us better understand what's going on**:
[Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.]

* `kubectl get pods -n openebs`
```
NAME                                                              READY   STATUS             RESTARTS   AGE
openebs-provisioner-c4c58475c-ffqzh                               1/1     Terminating        21         41d
pvc-f39179d6-2121-439d-ae18-bd4bd8f47f8d-target-78f9b46998xqt99   3/3     Running            0          38d
pvc-614627ce-bcf9-4d9e-941a-9408769679d7-target-5d4fb856b4x96qc   3/3     Running            0          38d
openebs-localpv-provisioner-848db54f69-98452                      1/1     Running            75         41d
pvc-1c329d55-f024-4306-8489-0efbd65f58c5-target-69bdf7d56-wp75s   3/3     Running            0          38d
openebs-ndm-8bt8x                                                 1/1     Running            1          46d
openebs-snapshot-operator-84fb87cf4c-q4f4v                        2/2     Running            46         41d
openebs-admission-server-6dbc7cbd85-wjrd8                         1/1     Running            0          41d
pvc-b72ec2de-e301-4667-92cc-43d089dd2e02-target-5f6447cdf5lk2zc   3/3     Running            0          38d
maya-apiserver-66d7fdc9bd-5h5wk                                   1/1     Running            0          41d
cstor-disk-pool-eiq0-7cfdc9844-mf7dc                              3/3     Running            0          5d18h
openebs-ndm-7cgqd                                                 1/1     Running            1          29d
pvc-318b2f41-40db-4ee7-8564-4c8677665bbf-target-55cb656944cgrz5   3/3     Running            0          19d
pvc-20bffc04-703d-44b0-8820-e124a9c0bc51-target-566b7f5f7fsff9b   3/3     Running            0          19d
openebs-ndm-8j7nw                                                 1/1     Running            10         46d
cstor-disk-pool-una0-74c98c99d7-ztr85                             3/3     Running            0          7d19h
pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-target-875dc5d79-7q8wg   3/3     Running            1          38d
openebs-ndm-wj5sx                                                 1/1     Running            4          41d
cstor-disk-pool-7it6-77c8db748f-w2qpg                             3/3     Running            0          11d
openebs-provisioner-c4c58475c-jl847                               1/1     Running            147        19d
openebs-ndm-operator-65bb6b586b-nlpxp                             0/1     CrashLoopBackOff   875        41d
```

* `kubectl get blockdevices -n openebs -o yaml`

Without `-o yaml`:
NAME                                           NODENAME   SIZE            CLAIMSTATE   STATUS     AGE
blockdevice-6c32c218adf505e3ccc8f0c4149ea788   karen      2000398934016   Unclaimed    Inactive   46d
blockdevice-bda470640c397d416b38c1cf3d4920a1   karen      0               Unclaimed    Inactive   46d
blockdevice-06966035bfc93af9d1ad00cfea38eaf7   karen      0               Unclaimed    Inactive   46d
blockdevice-0d6f165f287d6f18d52c7058a36b52fb   karen      2000396746752   Unclaimed    Inactive   46d
blockdevice-5f71eb771e5493870c4172c7ed350b70   karen      0               Unclaimed    Inactive   40d
blockdevice-1928111f2d55f6ca491487c5bbdb1cef   karen      2000396746752   Unclaimed    Inactive   41d
blockdevice-c6f33099b79733d452f4cdb230035fad   karen      2000396746752   Unclaimed    Active     46d
blockdevice-7fcbc73a36122ec29aa4332aa1e0fd34   kristel    0               Unclaimed    Inactive   31d
blockdevice-37662249ac2cba98ce6b4b6831ee1e31   josje      0               Unclaimed    Inactive   29d
blockdevice-9b31e9d5a1a302c25e887903d0a3944b   josje      0               Unclaimed    Inactive   29d
blockdevice-b1c545fd574fba219ff6e12140bea876   josje      42949672960     Unclaimed    Inactive   29d
blockdevice-b86b3018f7ae807a1d336665f0af2685   kristel    3000592982016   Claimed      Active     12d
blockdevice-55bebd17eca81d998467cd0cfeecab7e   karen      3000592982016   Claimed      Active     7d19h
blockdevice-cd0297590c2bc7be872ea0894f912be1   kathleen   500107862016    Unclaimed    Inactive   41d
blockdevice-9038ee5f84859d03e8c0ad2993a58bf8   josje      0               Unclaimed    Inactive   31d
blockdevice-3ce583a5ae11343272b9ae2a044798a0   kathleen   3000592982016   Unclaimed    Active     5d18h

There's quite a lot of blockdevices, so I'll post the YAML only for the three BDs that are actually in use, which are the three disks of size 3000592982016:

https://pastebin.com/yJEgjVMv

* `kubectl get blockdeviceclaims -n openebs -o yaml`

```
NAME                                       BLOCKDEVICENAME                                PHASE   AGE
bdc-eaeec1a7-ea04-4848-b2cd-b7c3c5afd643   blockdevice-b86b3018f7ae807a1d336665f0af2685   Bound   12d
bdc-2a2f7a1c-2e74-4253-887c-d286b2829e43   blockdevice-55bebd17eca81d998467cd0cfeecab7e   Bound   5d18h
bdc-82010b96-459f-46bf-b371-9ba6424d99fd   blockdevice-3ce583a5ae11343272b9ae2a044798a0   Bound   5d18h
bdc-ebf6da43-64da-4033-8c9b-82cefad4fc0b   blockdevice-e728c07ebb509b1ab5881e975f8189a2   Bound   46d
```

https://pastebin.com/E1wWkngN


**Environment:**
- OpenEBS version: 1.9.0 with patches to cStor to improve arm64 support
- Kubernetes version (use `kubectl version`): v1.17.4+k3s1
- Kubernetes installer & version: k3os
- Cloud provider or hardware configuration: arm64 devices (mix of Raspberry Pi 3B, Raspberry Pi 4 and Orange Pi PC)
- Type of disks connected to the nodes (eg: Virtual Disks, GCE/EBS Volumes, Physical drives etc): Physical drives via USB-SATA
- OS (e.g. from `/etc/os-release`): k3os v0.10.0

I just upgraded my control plane to 1.11.0 as it was released today. I haven't upgraded my data plane yet. The new pod is running `node-disk-operator` 0.6.0 now:

```
$ kubectl describe pod -n openebs openebs-ndm-operator-694c6c48d9-tgxsn | grep -i image
    Image:          quay.io/openebs/node-disk-operator-arm64:0.6.0
    Image ID:       quay.io/openebs/node-disk-operator-arm64@sha256:1aea4ad627c906b993d313e605c22f1ceefbb2feee3f3d73bbd2103875ed5329
      CLEANUP_JOB_IMAGE:  quay.io/openebs/linux-utils-arm64:1.11.0
  Normal   Pulled     25m (x4 over 28m)     kubelet, josje     Successfully pulled image "quay.io/openebs/node-disk-operator-arm64:0.6.0"
  Normal   Pulling    3m38s (x10 over 28m)  kubelet, josje     Pulling image "quay.io/openebs/node-disk-operator-arm64:0.6.0"
```

However, it is still crashing:

```
$ kubectl logs -n openebs openebs-ndm-operator-694c6c48d9-tgxsn
I0613 19:56:47.017241       1 main.go:52] Go Version: go1.14.1
I0613 19:56:47.018216       1 main.go:53] Go OS/Arch: linux/arm64
I0613 19:56:47.018311       1 main.go:54] operator-sdk Version: v0.5.0
I0613 19:56:47.018348       1 main.go:55] Version Tag: v0.6.0
I0613 19:56:47.018531       1 main.go:56] Git Commit: 7294b57baa2188b808eda30f9ee6e715c5f94097
{"level":"info","ts":1592078207.023273,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1592078207.408349,"logger":"leader","msg":"Found existing lock with my name. I was likely restarted."}
{"level":"info","ts":1592078207.4084928,"logger":"leader","msg":"Continuing as the leader."}
I0613 19:56:47.624864       1 main.go:112] Installing the components
I0613 19:56:55.830260       1 main.go:127] Registering Components
I0613 19:56:56.281529       1 main.go:142] Check if CR has to be upgraded, and perform upgrade
{"level":"info","ts":1592078216.3678615,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"blockdevice-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1592078216.3688335,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"blockdeviceclaim-controller","source":"kind source: /, Kind="}
I0613 19:56:56.369402       1 main.go:155] Starting the ndm-operator...
{"level":"info","ts":1592078216.4730475,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"blockdeviceclaim-controller"}
{"level":"info","ts":1592078216.4729877,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"blockdevice-controller"}
{"level":"info","ts":1592078216.5735834,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"blockdeviceclaim-controller","worker count":1}
{"level":"info","ts":1592078216.573484,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"blockdevice-controller","worker count":1}
I0613 19:56:56.574567       1 blockdeviceclaim_controller.go:262] Releasing claimed block device blockdevice-e728c07ebb509b1ab5881e975f8189a2 from bdc-ebf6da43-64da-4033-8c9b-82cefad4fc0b
E0613 19:56:56.575739       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/runtime/runtime.go:76
/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/runtime/runtime.go:65
/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/runtime/runtime.go:51
/home/travis/.gimme/versions/go1.14.1.linux.arm64/src/runtime/panic.go:967
/home/travis/.gimme/versions/go1.14.1.linux.arm64/src/runtime/panic.go:212
/home/travis/.gimme/versions/go1.14.1.linux.arm64/src/runtime/signal_unix.go:695
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:244
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:274
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:203
/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:128
/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:215
/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:158
/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:133
/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:134
/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:88
/home/travis/.gimme/versions/go1.14.1.linux.arm64/src/runtime/asm_arm64.s:1148
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xf497e4]

goroutine 189 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/runtime/runtime.go:58 +0x104
panic(0x10acc80, 0x1cc20e0)
	/home/travis/.gimme/versions/go1.14.1.linux.arm64/src/runtime/panic.go:967 +0x11c
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).isDeviceRequestedByThisDeviceClaim(...)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:244
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).releaseClaimedBlockDevice(0x400047f2c0, 0x40001da1a0, 0x1, 0x122dfd4)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:274 +0x3a4
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).FinalizerHandling(0x400047f2c0, 0x40001da1a0, 0x1228e26, 0x14)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:203 +0x98
github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim.(*ReconcileBlockDeviceClaim).Reconcile(0x400047f2c0, 0x4000524540, 0x7, 0x4000047920, 0x28, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/openebs/node-disk-manager/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go:128 +0x3b8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x400057e280, 0x400005a700)
	/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:215 +0x154
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:158 +0x34
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0x40001a39f0)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:133 +0x64
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x40001a39f0, 0x3b9aca00, 0x0, 0x1, 0x4000545c20)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:134 +0xb0
k8s.io/apimachinery/pkg/util/wait.Until(0x40001a39f0, 0x3b9aca00, 0x4000545c20)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:88 +0x48
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:157 +0x2d0
```

Should I also upgrade the data plane before attempting this again?

--- a/changelogs/unreleased/479-akhilerm
+++ b/changelogs/unreleased/479-akhilerm
@@ -1,0 +1,1 @@
+fix a bug where NDM operator crashes if a claimed BD is manually deleted

--- a/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go
+++ b/pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go
@@ -298,6 +298,15 @@ func (r *ReconcileBlockDeviceClaim) releaseClaimedBlockDevice(
 			break
 		}
 	}
+	// This case occurs when a claimed BD is manually deleted by removing the finalizer.
+	// If this check is not performed, the NDM operator will continuously crash, because it
+	// will try to release a non existent BD.
+	if claimedBd == nil {
+		r.recorder.Eventf(instance, corev1.EventTypeWarning, "BlockDeviceNotFound", "BlockDevice %s not found for releasing", instance.Spec.BlockDeviceName)
+		klog.Errorf("could not find blockdevice for claim: %s", instance.Name)
+		return fmt.Errorf("blockdevice: %s not found for releasing from bdc: %s", instance.Spec.BlockDeviceName, instance.Name)
+	}
+
 	dvr := claimedBd.DeepCopy()
 	dvr.Spec.ClaimRef = nil
 	dvr.Status.ClaimState = apis.BlockDeviceReleased

